lib/httplib/connstate.go

```diff
@@ -0,0 +1,237 @@
+		c = NewConnStateEvents()
+		s.conns[conn] = c
+	}
+	return c.Channel(state)
+}
+
+// Release resources for conn when something goes wrong after calling
+// Channel and before http.Serve registers conn to trigger events. This
+// will remove net.Conn state. This method is only required if http.Serve
+// returns an error because state is removed when connections enter a
+// terminal state.
+//
+// The conn will be tracked again, without previously registered
+// subscriptions, if it is passed to Notify after calling Release.
+func (s *ServerConnState) Release(conn net.Conn) {
+	s.mu.Lock()
+	c := s.conns[conn]
+	delete(s.conns, conn)
+	s.mu.Unlock()
+
+	if c != nil {
+		c.Stop()
+	}
+}
+
+// ConnStateEvents tracks connection state for a single connection.
+// This type is created by ServerConnState for each distinct connection.
+type ConnStateEvents struct {
+	events chan interface{}
+}
+
+// NewConnStateEvents returns an initialized ConnStateEvents.
+func NewConnStateEvents() *ConnStateEvents {
+	c := &ConnStateEvents{events: make(chan interface{}, connStateEventsQueueSize)}
+	go c.processEvents()
+	return c
+}
+
+// Notify state change.
+func (c *ConnStateEvents) Notify(state http.ConnState) {
+	c.events <- state
+}
+
+// Channel returns a read-only channel that will receive http.ConnState events.
+// The state parameter determines which states are received by the channel.
+// The channel will receive all events when state is AllConnStates. All channels
+// will receive terminal states (http.StateClosed & http.StateHijacked) events
+// after which the channel is closed. Multiple calls are supported.
+//
+// Passing a closed connection will result in no events received and leak
+// the channel unless followed by a Stop call.
+//
+// Calling Channel when already in the desired state will result
+// in no events until state moves to a different state and back
+// again (e.g. idle->active->idle).
+func (c *ConnStateEvents) Channel(state http.ConnState) <-chan http.ConnState {
+	// chan buffer=1: assume channel is being serviced, but don't sync with reader
+	ch := make(chan http.ConnState, 1)
+	c.events <- connStateChannel{state: state, ch: ch}
+	return ch
+}
+
+// Stop receiving events and close all subscriber channels.
+func (c *ConnStateEvents) Stop() {
+	c.events <- connStateStop{}
+}
+
+// processEvents is the core goroutine that runs for each instance.
+// It serializes all events: channel subscriptions, state change events,
+// and stop requests. It exits when it receives a connStateStop message,
+// which can be accomplished by calling Stop.
+func (c *ConnStateEvents) processEvents() {
+	var channels []connStateChannel
+	for event := range c.events {
+		switch v := event.(type) {
+		case connStateChannel:
+			channels = append(channels, v)
+		case http.ConnState:
+			c.notify(channels, v)
+			if TerminalConnState(v) {
+				return
+			}
+		case connStateStop:
+			for _, channel := range channels {
+				close(channel.ch)
+			}
+			return
+		}
```

Comment:
Add a `default` case and log a warning if received an unknown event type?

Reply:
I don't currently have access to a log in this object. I thought of it as more of a library. All of the message types are generated by itself, so I thought it was safe to not have a default. LMK if you want me to add logging. My feeling is it isn't needed, but it would be simple to add.

Reply:
Yeah makes sense, it's fine as it is.